### PR TITLE
DSP-676 Add Support for Additional Methods in Admin Permission Endpoint 

### PR DESCRIPTION
**Pull request by tobiasschweizer**
resolves [DSP-676](https://dasch.myjetbrains.com/youtrack/issue/DSP-676)

### CODE REVIEW

**Comment by tobiasschweizer** [comment]
## A Project's permissions

Route: `GET: /admin/permissions/<projectIri>`
Method: getPermissions

Created classes: `ProjectPermissionsResponse` and `ProjectPermission`

## Administrative Permissions

### Get Administrative Permissions

Route: `GET: /admin/permissions/ap/<projectIri>`
Method: `getAdministrativePermissions`

Created class: `AdministrativePermissionsResponse`

Changed class: added "iri" to `AdministrativePermission` (reading)

### Create Administrative Permission

Route: `POST: /admin/permissions/ap`
Method: `createAdministrativePermission`

Created class: `CreateAdministrativePermission` (with optional "id")

**Comment by tobiasschweizer** [comment]
## Open Questions

### Relation between DOAP and AP

I made `CreateDefaultObjectAccessPermission` a subclass of `CreateAdministrativePermission` because default object access permission can reuse the class `AdministrativePermission`. `DefaultObjectAccessPermissionResponse` also uses the class `AdministrativePermission`. Does it make sense to think of a DOAP as a AP?

### Which Properties are Optional / "Nullable"?

- `CreatePermission`: `additionalInformation` and `permissionCode` can be `null`, `name` is required
- `CreateAdministrativePermission`:

**Comment by tobiasschweizer** [comment]
After discussion with @subotic 

- administrative permissions and default object access permissions are different concepts and should be represented by different classes both on reading and creating. 
- for the moment, we have to provide different classes for reading and writing because of nullable props (`json2typescript`) and because of the inconsistency between id and iri.

**Comment by tobiasschweizer** [comment]
## Reading Permissions

base class:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/admin-doap-base.ts#L4-L25

administrative permission:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/administrative-permission.ts#L7-L16

object access default permission:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/default-object-access-permission.ts#L4-L25

**Comment by tobiasschweizer** [comment]
## Creating Permissions

base class:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/create-admin-doap-base.ts#L4-L25

administrative permission:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/create-administrative-permission.ts#L8-L17

object access default permission:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/create-default-object-access-permission.ts#L7-L28

**Comment by tobiasschweizer** [comment]
@subotic I also found `StoredAdministrativePermission`:

https://github.com/dasch-swiss/dsp-js-lib/blob/50989aeebc80b019ca92b77ab496c93d0e869e63/src/models/admin/stored-administrative-permission.ts#L8-L17

It is not used anywhere. And since we have separate classes for creating and writing anyway, I think we could get rid of it.

File: src/api/admin/permissions/permissions-endpoint-admin.ts [hunk 1]
```diff
@@ -1,18 +1,43 @@
...
-import { AdministrativePermissionResponse } from "../../../models/admin/administrative-permission-response";
-
-
 /**
  * An endpoint for working with Knora permissions.
  */
 export class PermissionsEndpointAdmin extends Endpoint {
-    
+
+    getPermissions(projectIri: string): Observable<ApiResponseError | ApiResponseData<ProjectPermissionsResponse>> {
```
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts:24]: I think you are right, I will rename this method.

[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts]: done in https://github.com/dasch-swiss/dsp-js-lib/pull/243/commits/83d33e4460c1672070f5b969bc6f4261e5874967

File: src/api/admin/permissions/permissions-endpoint.spec.ts [hunk 2]
```diff
@@ -18,9 +25,75 @@ describe("PermissionsEndpoint", () => {
         jasmine.Ajax.uninstall();
     });
 
+    describe("Method getPermissions", () => {
```
Comment (kilchenmann) [inline src/api/admin/permissions/permissions-endpoint.spec.ts:28]: "Method getProjectPermissions"

**Review by kilchenmann — APPROVED** [review]

File: src/api/admin/permissions/permissions-endpoint.spec.ts (hunk 2 above)
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint.spec.ts]: good catch

fixed in https://github.com/dasch-swiss/dsp-js-lib/pull/243/commits/5cb4977e9b968f402ff068a8f95743c9c6b3d2d5

**Comment by kilchenmann** [comment]
@tobiasschweizer Can you write PR titles as described in the ["Contribution" documentation](https://docs.dasch.swiss/developers/dsp/contribution/#pull-request-guidelines):

`<DSP-nr> <title>`

Thanks

**Comment by tobiasschweizer** [comment]
> @tobiasschweizer Can you write PR titles as described in the ["Contribution" documentation](https://docs.dasch.swiss/developers/dsp/contribution/#pull-request-guidelines):
> 
> `<DSP-nr> <title>`
> 
> Thanks

ok, done

**Comment by tobiasschweizer** [comment]
@subotic Could you please have a look at this PR? Thanks!

**Review by subotic — APPROVED** [review]
Cool. There are two small things.
On a side-note. Like the App for E2E testing. I immediately understood what is happening and how I could extend it 👍 .

File: src/api/admin/permissions/permissions-endpoint-admin.ts [hunk 3]
```diff
@@ -27,5 +57,78 @@ export class PermissionsEndpointAdmin extends Endpoint {
...
+     *
+     * @param administrativePermission the administrative permission to be created.
+     */
+    createAdministrativePermission(administrativePermission: CreateAdministrativePermission): Observable<ApiResponseError | ApiResponseData<AdministrativePermissionResponse>> {
+
+        if (!administrativePermission.forGroup || !administrativePermission.forProject) {
+            throw new Error("Group and project are required when creating a new administrative permission.");
+        }
+
+        if (administrativePermission.hasPermissions.length !== 1) {
```
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts:72]: Would the response then contain an array of permissions?

The problem is that then the response format would be different depending on how many permissions were created:
- one permission: "administrative_permission"
- more than one permission: "administrative_permissions"

If I understand correctly, the test data covers the creation of exactly one permission.

File: src/api/admin/permissions/permissions-endpoint-admin.ts [hunk 4]
```diff
@@ -27,5 +57,78 @@ export class PermissionsEndpointAdmin extends Endpoint {
...
+     */
+    createDefaultObjectAccessPermission(defaultObjectAccessPermission: CreateDefaultObjectAccessPermission): Observable<ApiResponseError | ApiResponseData<DefaultObjectAccessPermissionResponse>> {
+
+        // A default object access permission must
+        // always reference a project
+        if (!defaultObjectAccessPermission.forProject) {
+            throw new Error("Project is required when creating a new default object access permission.");
+        }
+
+        if (defaultObjectAccessPermission.hasPermissions.length !== 1) {
```
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts:110]: please see https://github.com/dasch-swiss/dsp-js-lib/pull/243#discussion_r501663873

File: src/api/admin/permissions/permissions-endpoint-admin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts:72]: Ok, I think I confused something. I tried to submit several permissions in the array and it worked: :-)

The answer is still "administrative_permission"

[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts]: now the method just checks that there is at least one permission: https://github.com/dasch-swiss/dsp-js-lib/pull/243/commits/f6bbdef9ef8a323dea2ffc53f3eed051cdbf9184

File: src/api/admin/permissions/permissions-endpoint-admin.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tobiasschweizer) [reply src/api/admin/permissions/permissions-endpoint-admin.ts]: now the method just checks that there is at least one permission: f6bbdef

File: src/api/admin/permissions/permissions-endpoint-admin.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (subotic) [reply src/api/admin/permissions/permissions-endpoint-admin.ts]: great, thanks.

**Review by subotic — APPROVED** [review]

**Comment by subotic** [comment]
> @tobiasschweizer Can you write PR titles as described in the ["Contribution" documentation](https://docs.dasch.swiss/developers/dsp/contribution/#pull-request-guidelines):
> 
> `<DSP-nr> <title>`
> 
> Thanks

@kilchenmann I guess that the title should be the same as in Youtrack? Maybe this could be explicitly mentioned in the guidelines?

**Comment by kilchenmann** [comment]
> I guess that the title should be the same as in Youtrack?

@subotic Good point

**Pull request merged**